Unit Tests for API

sabinevidal opened the pull request:
@Ashara-kosi

lacabra commented:
@sabinevidal I believe the easiest would be for you to add @Ashara-kosi as a contributor to your forked repo, so that she can push contributions to feature branch `api-unit-tests` on your repo, which will automatically appear on this PR.

lacabra commented on api/api.py:
```diff
@@ -105,11 +105,11 @@ def process(inputText):
 
     # PT
-    model = AutoModelForSequenceClassification.from_pretrained('./model')
+    model = AutoModelForSequenceClassification.from_pretrained('./api/model')
```
While this is a possible fix, I would ask not to change this path because this will break other dependencies. The root cause of this path mismatch lies elsewhere, revert this back to `./model`, and instead move the `test_kindly.py` file inside the `api/` folder so that this relative path will work properly. Plus this repository contains various different components written in different languages (Python vs NodeJS), so it's good to keep things organized, and we are keeping all Python code inside the `api/` folder. While you are moving this file, rename it to `test_api.py` as it is a more specific name to the component you are testing.

lacabra commented on api/api.py:
```diff
@@ -105,11 +105,11 @@ def process(inputText):
     text = inputText
     text = preprocess(text)
-    tokenizer = AutoTokenizer.from_pretrained('./model')
+    tokenizer = AutoTokenizer.from_pretrained('./api/model')
```
Same as above

lacabra commented on api/requirements.txt:
```diff
@@ -10,6 +10,7 @@ matplotlib==3.4.2
 mwparserfromhell==0.6.2
 numpy==1.19.5
 pandas==1.3.1
+pytest==6.2.5
```
Please add this same line to `requirements.python-3.6.8.txt` to keep both files in sync, thank you!

lacabra commented on docs/development.md:
```diff
@@ -173,3 +173,17 @@ axios(config)
+
+Unit test for endpoints have been set up using pytest.
+
+To run tests from main directory:
```
Edit the instructions to run the tests from inside the `api/` folder as per the comment above

lacabra commented on test_kindly.py:
```diff
@@ -0,0 +1,53 @@
+    """ Testing 403 error if no correct headers"""
+    offensive_text = {'text': 'test'}
+    with app.test_client() as c:
+        response = c.post('/detect', json=offensive_text, base_url = 'http://localhost:8080/')
```
Please note that you are hitting a different server at port `8080`. You should run all your test against the configured server at port `3000`

lacabra commented on test_kindly.py:
```diff
@@ -0,0 +1,53 @@
+from api.api import app, checkHeaders, detect
+
+app.testing = True
+headers = {'Authorization': 'Bearer aasdf1234', 'Content-Type': 'application/json'}
```
This assumes that `.env` is setup in some particular way in your local environment, but the tests should be run independently of any local environment configuration. Can you please instead look into setting up the `pytest-env` plugin, and configuring there the specific environment variables that you need?

sabinevidal commented:
@lacabra 
Have made requested changes and added more endpoint tests.
Notes:
- used `pytest-dotenv` as `pytest-env` is unmaintained
- the `/train` endpoint seems to be incomplete as `NameError: name 'emotion' is not defined` is thrown when trying to access. Have added one test for it, but have commented it out.
- Added tests for expected errors currently in `api.py` for each endpoint ie `400` and `403` and one generic `403`.
